google/resumable_media/requests/download.py

```diff
@@ -58,12 +62,43 @@ def _write_to_stream(self, response):
 
         Args:
             response (~requests.Response): The HTTP response object.
+
+        Raises:
+            DataCorruption: If the download's checksum doesn't agree with
+                server-computed checksum.
         """
+        md5_hash = hashlib.md5()
+        # TODO: Add support for validating CRC32C for composite objects. Note
+        # that this is somewhat painful in Python 2 because most OS distros of
+        # Python 2 don't include a compiled crcmod (and executing without a
+        # compiled crcmod is very slow; and installing a compiled crcmod will
+        # make getting google-cloud-python working well more difficult).
+        expected_md5_hash = None
+        if ('X-Goog-Hash' in response.headers
+            and response.headers['X-Goog-Hash']):
```

Comment:
First, this should go in the base module `google.resumable_media._download` rather than the `requests`-specific implementation.

----

I suggest instead:

```python
headers = self._get_headers(response)
goog_hash = headers.get('X-Goog-Hash', '')
if goog_hash:
    ...
```

Are you worried about `headers` being case-sensitive (it won't be for `requests`)?

Reply:
google.resumable_media._download doesn't have code that walks through the content chunk-by-chunk. So, for example, if I put the MD5 checksumming in _process_response it would either get the entire body at once (using self._get_body(response)) or I would have to add another loop over response.iter_content that walks through the chunks. The former would require too much memory for large downloads, and the latter would mean we're walking over the content twice (once in google.resumable_media._download, and once in google.resumable_media.requests.download, in _write_to_stream) - and I think the latter would fail because iter_content walks through the HTTP stream, and it wouldn't be available twice (well, not without starting a separate connection to fetch the content again). Both approaches seem much worse than what I have right now, but maybe I'm not understanding something about what you suggest?